L1Trigger/L1THGCal/src/be_algorithms/HGCalShowerShape.cc

```diff
@@ -0,0 +1,554 @@
+
+    const edm::PtrVector<l1t::HGCalCluster> clustersPtrs = c3d.constituents();
+
+    int firstLayer=999;
+    int lastLayer=-999;
+    int layer=999;
+
+    int nLayersEE=28;
+   
+    for(const auto& clu : clustersPtrs){
+
+        if(clu->subdetId()==HGCEE) layer=clu->layer();//EE
+        if(clu->subdetId()==HGCHEF) layer=clu->layer()+nLayersEE;//FH
+
+        if(layer<firstLayer) firstLayer=layer;
+        if(layer>lastLayer) lastLayer=layer;
+
+    }
+
+    int nLayers=lastLayer-firstLayer+1;
+
+    return nLayers;
+
+}
+
+
+const int HGCalShowerShape::firstLayer(const l1t::HGCalMulticluster& c3d){
+
+    const edm::PtrVector<l1t::HGCalCluster> clustersPtrs = c3d.constituents();
+
+    int firstLayer=999;
+    int layer=999;
+
+    int nLayersEE=28;
+
+    for(const auto& clu : clustersPtrs){
+
+        if(clu->subdetId()==HGCEE) layer=clu->layer();//EE
+        if(clu->subdetId()==HGCHEF) layer=clu->layer()+nLayersEE;//FH
+
+        if(layer<firstLayer) firstLayer=layer;
+
+    }
+
+    return firstLayer;
+
+}
+
+
+
+
+const float HGCalShowerShape::sigmaEtaEtaTot(const l1t::HGCalMulticluster& c3d){
+
+    const edm::PtrVector<l1t::HGCalCluster> clustersPtrs = c3d.constituents();
+
+    std::vector<float> tc_energy ; 
+    std::vector<float> tc_eta ;
+
+    for(const auto& clu : clustersPtrs){
+
+        const edm::PtrVector<l1t::HGCalTriggerCell> triggerCells = clu->constituents();
+        unsigned int ncells = triggerCells.size();
+
+        for(unsigned int itc=0; itc<ncells;itc++){
+
+            l1t::HGCalTriggerCell thistc = *triggerCells[itc];
+
+            tc_energy.emplace_back(thistc.energy());
+            tc_eta.emplace_back(thistc.eta());
+                
+        }
+
+    }
+
+    float SeeTot = sigmaEtaEta(tc_energy,tc_eta);
+
+    tc_energy.clear();
+    tc_eta.clear();
+
+    return SeeTot;
+
+}
+
+
+
+
+const float HGCalShowerShape::sigmaPhiPhiTot(const l1t::HGCalMulticluster& c3d){
+
+
+    const edm::PtrVector<l1t::HGCalCluster> clustersPtrs = c3d.constituents();
+
+    std::vector<float> tc_energy ; 
+    std::vector<float> tc_phi ;
+
+    for(const auto& clu : clustersPtrs){
+
+        const edm::PtrVector<l1t::HGCalTriggerCell> triggerCells = clu->constituents();
+        unsigned int ncells = triggerCells.size();
+
+        for(unsigned int itc=0; itc<ncells;itc++){
+
+            l1t::HGCalTriggerCell thistc = *triggerCells[itc];
+
+            tc_energy.emplace_back(thistc.energy());
+            tc_phi.emplace_back(thistc.phi());
+                
+        }
+    }
+
+    float SppTot = sigmaPhiPhi(tc_energy,tc_phi);
+
+    tc_energy.clear();
+    tc_phi.clear();
+
+    return SppTot;
+
+}
+
+
+
+const float HGCalShowerShape::sigmaEtaEtaMax(const l1t::HGCalMulticluster& c3d){
+
+    const edm::PtrVector<l1t::HGCalCluster> clustersPtrs = c3d.constituents();
+
+    std::vector<int> layer ; // Size : ncl2D
+    std::vector<int> subdetID ;
+    std::vector<int> nTC ;
+    std::vector<float> tc_energy ; 
```

Comment:
Why not using a vector for each layer, since later the shapes are computed per layer?